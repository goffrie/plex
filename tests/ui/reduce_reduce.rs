//! This test demonstrates a reduce-reduce conflict.

enum Token {
    Lit(i32),
    Var(i32),
    Minus,
}
type Span = ();
enum Expr {
    Lit(i32),
    Var(i32),
    Negate(Box<Expr>),
}
#[allow(unused_imports)]
use self::Token::*;
plex::parser! {
    fn parse_(Token, Span);

    expr: Expr {
        Var(i) => Expr::Var(i),
        Lit(i) => Expr::Lit(i),
        // These rules induce a reduce-reduce conflict.
        // The intent here is to create a special case where `Minus` followed by a `Lit` is parsed as a negative literal,
        // rather than a negation expression.
        // To implement that, use the attribute:
        // #[overriding]
        Minus Lit(i) => Expr::Lit(-i),
        Minus expr[e] => Expr::Negate(Box::new(e)),
    }
}

fn main() {}
