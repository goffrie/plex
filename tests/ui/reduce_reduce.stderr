error: reduce-reduce conflict:
       state: expr -> Lit •
              expr -> Minus Lit •
       token: EOF
  --> tests/ui/reduce_reduce.rs:27:9
   |
27 |         Minus Lit(i) => Expr::Lit(-i),
   |         ^^^^^^^^^^^^^^^

error: this is the conflicting rule
  --> tests/ui/reduce_reduce.rs:21:9
   |
21 |         Lit(i) => Expr::Lit(i),
   |         ^^^^^^^^^
